gossip: TestConvergence times out: not connected to cluster; use --join to specify a connected node

Happened locally on WIP (which did not touch gossip), on top of e4cdd4503104811ce082122ce6bc5adf392f9b31.

```
I160405 15:55:52.981211 gossip/simulation/network.go:57  simulating gossip network with 10 nodes
W160405 15:55:52.981326 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981365 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981472 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981501 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.981532 gossip/gossip.go:917  starting client to 127.0.0.1:50262
W160405 15:55:52.981593 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981619 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.981652 gossip/gossip.go:917  starting client to 127.0.0.1:50262
W160405 15:55:52.981709 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981737 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.981773 gossip/gossip.go:917  starting client to 127.0.0.1:50262
W160405 15:55:52.981828 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.981840 gossip/gossip.go:917  starting client to 127.0.0.1:50262
W160405 15:55:52.981855 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981945 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.981981 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.982029 gossip/gossip.go:917  starting client to 127.0.0.1:50262
W160405 15:55:52.982124 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.982160 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.982213 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:52.982252 gossip/gossip.go:917  starting client to 127.0.0.1:50262
W160405 15:55:52.982258 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.982288 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.982392 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.982423 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.982505 gossip/server.go:182  refusing gossip from node 4 (max 3 conns); forwarding to 2 ({tcp 127.0.0.1:50263})
W160405 15:55:52.982554 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.982585 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.982611 gossip/server.go:182  refusing gossip from node 7 (max 3 conns); forwarding to 2 ({tcp 127.0.0.1:50263})
I160405 15:55:52.982625 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:52.982666 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:52.982692 gossip/server.go:182  refusing gossip from node 5 (max 3 conns); forwarding to 1 ({tcp 127.0.0.1:50262})
I160405 15:55:52.982739 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:52.982763 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:168: received forward from node 1 to 2 (127.0.0.1:50263)
I160405 15:55:52.982773 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:178: stopping outgoing client to node 1 (127.0.0.1:50262); loopback connection
W160405 15:55:52.982795 gossip/gossip.go:893  first range unavailable or cluster not initialized
I160405 15:55:52.982810 gossip/gossip.go:917  starting client to 127.0.0.1:50263
W160405 15:55:52.982820 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.982832 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:168: received forward from node 1 to 2 (127.0.0.1:50263)
I160405 15:55:52.982861 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:159: received forward from node 1 to 1 (127.0.0.1:50262); already have active connection, skipping
I160405 15:55:52.982869 gossip/server.go:182  refusing gossip from node 6 (max 3 conns); forwarding to 1 ({tcp 127.0.0.1:50262})
I160405 15:55:52.982880 gossip/gossip.go:917  starting client to 127.0.0.1:50263
W160405 15:55:52.982886 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:52.982911 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.982976 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:159: received forward from node 1 to 1 (127.0.0.1:50262); already have active connection, skipping
W160405 15:55:52.983035 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.983814 gossip/server.go:182  refusing gossip from node 8 (max 3 conns); forwarding to 2 ({tcp 127.0.0.1:50263})
I160405 15:55:52.984110 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:168: received forward from node 1 to 2 (127.0.0.1:50263)
I160405 15:55:52.984355 gossip/gossip.go:917  starting client to 127.0.0.1:50263
I160405 15:55:52.984485 gossip/server.go:182  refusing gossip from node 10 (max 3 conns); forwarding to 1 ({tcp 127.0.0.1:50262})
I160405 15:55:52.984561 gossip/server.go:182  refusing gossip from node 9 (max 3 conns); forwarding to 1 ({tcp 127.0.0.1:50262})
I160405 15:55:52.984765 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:159: received forward from node 1 to 1 (127.0.0.1:50262); already have active connection, skipping
W160405 15:55:52.984790 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:52.984852 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:159: received forward from node 1 to 1 (127.0.0.1:50262); already have active connection, skipping
I160405 15:55:53.981655 gossip/gossip.go:917  starting client to 127.0.0.1:50263
I160405 15:55:53.982077 gossip/server.go:182  refusing gossip from node 1 (max 3 conns); forwarding to 8 ({tcp 127.0.0.1:50270})
I160405 15:55:53.982109 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:53.982131 gossip/client.go:84  closing client to node 2 (127.0.0.1:50263): gossip/client.go:182: stopping outgoing client to node 2 (127.0.0.1:50263); already have incoming
I160405 15:55:53.982320 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:53.982571 gossip/server.go:182  refusing gossip from node 6 (max 3 conns); forwarding to 5 ({tcp 127.0.0.1:50267})
I160405 15:55:53.982698 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:168: received forward from node 1 to 5 (127.0.0.1:50267)
I160405 15:55:53.982733 gossip/gossip.go:917  starting client to 127.0.0.1:50267
W160405 15:55:53.982763 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
I160405 15:55:53.982813 gossip/gossip.go:917  starting client to 127.0.0.1:50262
I160405 15:55:53.983063 gossip/server.go:182  refusing gossip from node 10 (max 3 conns); forwarding to 3 ({tcp 127.0.0.1:50264})
I160405 15:55:53.983183 gossip/client.go:84  closing client to node 1 (127.0.0.1:50262): gossip/client.go:168: received forward from node 1 to 3 (127.0.0.1:50264)
I160405 15:55:53.983216 gossip/gossip.go:917  starting client to 127.0.0.1:50264
W160405 15:55:54.982870 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:55.984137 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:56.984254 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:57.984346 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:58.984414 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:55:59.984473 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:00.986059 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:01.986328 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:02.986866 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:03.987434 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:04.988574 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:06.001703 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:07.002961 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:08.003444 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:09.003591 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:10.003694 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:11.003781 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:12.003977 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:13.005008 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:14.006725 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:15.007222 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:16.007574 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:17.008273 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:18.008379 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:18.082760 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:19.009195 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:20.010342 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:21.010456 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:22.010853 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:23.011035 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:24.011651 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:25.011767 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:26.012257 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:27.012818 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:28.013253 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:29.013818 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:30.014422 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:31.019273 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:32.020271 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:33.020410 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:34.020491 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:35.021677 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:36.021972 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:37.023064 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:38.024056 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:39.024742 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:40.024984 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:41.025433 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:42.025557 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:43.025707 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:43.180882 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:44.026257 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:45.026365 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:46.027290 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:47.027560 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:48.028688 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:49.030131 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:50.030664 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:51.031150 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:52.031981 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:53.032800 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:54.033547 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:55.033751 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:56.033843 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:57.034436 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:58.035643 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:56:59.035972 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:57:00.036520 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:57:01.036639 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
W160405 15:57:02.037130 gossip/gossip.go:891  not connected to cluster; use --join to specify a connected node
panic: test timed out after 1m10s

goroutine 599616 [running]:
panic(0x461ff00, 0xc820449780)
	/usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6
testing.startAlarm.func1()
	/usr/local/Cellar/go/1.6/libexec/src/testing/testing.go:725 +0x14b
created by time.goFunc
	/usr/local/Cellar/go/1.6/libexec/src/time/sleep.go:129 +0x3a
```

full log: https://gist.github.com/tschottdorf/7c771ad96068c9fe3a70511671e1b4dd

--- a/gossip/server.go
+++ b/gossip/server.go
@@ -170,12 +170,23 @@ func (s *server) gossipReceiver(argsPtr **Request, senderFn func(*Response) erro
 				// Choose a random peer for forwarding.
 				altIdx := rand.Intn(len(s.nodeMap))
 				for addr, id := range s.nodeMap {
+					if id == s.is.NodeID {
+						// Don't forward back to ourselves.
+						continue
+					}
+					// Keep track of a valid forwarding peer in case the randomly
+					// selected node is the last node in the map and that node is
+					// ourself.
+					alternateAddr = addr
+					alternateNodeID = id
 					if altIdx == 0 {
-						alternateAddr = addr
-						alternateNodeID = id
 						break
 					}
 					altIdx--
+				}
+
+				if alternateNodeID == s.is.NodeID {
+					panic("cannot recommend self as node to forward to")
 				}
 
 				log.Infof("refusing gossip from node %d (max %d conns); forwarding to %d (%s)",

